Simplify server certificate and known hosts management

jannfis opened the pull request:
This PR aims to make it more simple to connect SSH and HTTPS repositories to ArgoCD. It addresses #1514 and the following topics:

- Make it more easy to add SSH host keys to the list of known hosts across all ArgoCD instances. This is accomplished by outsourcing the known hosts data into a ConfigMap, which in turn gets mounted as volume to the pods. There is no management interface yet, neither via CLI nor via UI, so to add new host keys one has to edit the ConfigMap (argocd-runtime-cm).
- Make it possible to connect HTTPS repositories with self-signed certificates, or those signed by unknown CAs. This PR introduces a new switch, "--insecure-skip-server-validation", which should replace the current "--insecure-ignore-host-key" switch. The current switch is kept for backwards compatibilty, but can possibly be removed.
- When adding repositories via CLI, the client now lets the server test the connection instead of performing a connection to the repository itself. The server now exposes a verification endpoint.

A management interface for SSH known hosts management has yet to be introduced, as well as an interface for managing HTTPS certificates.

jannfis commented:
This PR is work in progress. I have yet to write tests and polish the change. And since the change is quite substantial, I want to know your thoughts on the general direction this is going.

alexec commented:
looks like you're making good progress - let us know if you need any help

jessesuen commented:
> This PR introduces a new switch, "--insecure-skip-server-validation", which should replace the current "--insecure-ignore-host-key" switch. The current switch is kept for backwards compatibilty, but can possibly be removed.

So is this PR basically munging both the disabling of (a) TLS host certificate verification and (b) SSH host key verification into a single option `--insecure-skip-server-validation`?

I think I'm on-board with this change, however, I think the flag should be named `insecure-skip-server-verification` instead of `validation`.

alexec commented on manifests/ha/install.yaml:
```diff
@@ -1964,6 +1964,23 @@ metadata:
 apiVersion: v1
+data:
+  ssh_known_hosts: |
+    bitbucket.org ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEAubiN81eDcafrgMeLzaFPsw2kNvEcqTKl/VqLat/MaB33pZy0y3rJZtnqwR2qOOvbwKZYKiEO1O6VqNEBxKvJJelCq0dTXWT5pbO2gDXC6h6QDXCaHo6pOHGPUy+YBaGQRGuSusMEASYiWunYN0vCAI8QaXnWMXNMdFP3jHAJH0eDsoiGnLPBlBp4TNm6rYI74nMzgz3B9IikW4WVK+dc8KZJZWYjAuORU3jc1c/NPskD2ASinf8v3xnfXeukU0sJ5N6m5E8VLjObPEO+mN2t/FZTMZLiFqPWc/ALSqnMnnhwrNi2rbfg/rd/IpL8Le3pSBne8+seeFVBoGqzHM9yXw==
```
need to double check these obvs

alexec replied:
generated code

alexec commented on manifests/ha/namespace-install.yaml:
```diff
@@ -1879,6 +1879,23 @@ metadata:
 apiVersion: v1
+data:
+  ssh_known_hosts: |
+    bitbucket.org ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEAubiN81eDcafrgMeLzaFPsw2kNvEcqTKl/VqLat/MaB33pZy0y3rJZtnqwR2qOOvbwKZYKiEO1O6VqNEBxKvJJelCq0dTXWT5pbO2gDXC6h6QDXCaHo6pOHGPUy+YBaGQRGuSusMEASYiWunYN0vCAI8QaXnWMXNMdFP3jHAJH0eDsoiGnLPBlBp4TNm6rYI74nMzgz3B9IikW4WVK+dc8KZJZWYjAuORU3jc1c/NPskD2ASinf8v3xnfXeukU0sJ5N6m5E8VLjObPEO+mN2t/FZTMZLiFqPWc/ALSqnMnnhwrNi2rbfg/rd/IpL8Le3pSBne8+seeFVBoGqzHM9yXw==
```
as above

alexec replied:
generated code

alexec commented on manifests/install.yaml:
```diff
@@ -1822,6 +1822,23 @@ metadata:
 apiVersion: v1
+data:
+  ssh_known_hosts: |
+    bitbucket.org ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEAubiN81eDcafrgMeLzaFPsw2kNvEcqTKl/VqLat/MaB33pZy0y3rJZtnqwR2qOOvbwKZYKiEO1O6VqNEBxKvJJelCq0dTXWT5pbO2gDXC6h6QDXCaHo6pOHGPUy+YBaGQRGuSusMEASYiWunYN0vCAI8QaXnWMXNMdFP3jHAJH0eDsoiGnLPBlBp4TNm6rYI74nMzgz3B9IikW4WVK+dc8KZJZWYjAuORU3jc1c/NPskD2ASinf8v3xnfXeukU0sJ5N6m5E8VLjObPEO+mN2t/FZTMZLiFqPWc/ALSqnMnnhwrNi2rbfg/rd/IpL8Le3pSBne8+seeFVBoGqzHM9yXw==
```
reviewer to double check

alexec replied:
generated code

alexec commented on server/repository/repository.go:
```diff
@@ -82,7 +82,16 @@ func (s *Server) List(ctx context.Context, q *repositorypkg.RepoQuery) (*appsv1.
+			}
+			username := repo.Username
+			if username == "" {
+				username = "-"
```
um, can I ask what this is please?

jannfis replied:
This is supposed to have the username actually displayed in CLI's "repo list" function (the username never made its way back to the client).

I already moved the display logic to the client, but not pushed the change to my branch yet.

jannfis commented:
> So is this PR basically munging both the disabling of (a) TLS host certificate verification and (b) SSH host key verification into a single option `--insecure-skip-server-validation`?
> 
> I think I'm on-board with this change, however, I think the flag should be named `insecure-skip-server-verification` instead of `validation`.

@jessesuen Yes, correct, I think it would be best to provide a single option for both cases. I am with you that `verification` is a better term than `validation` for this process, I will change the name of the switch and also the code symbols accordingly.

I'm also in the process to extend the API to provide some kind of rudimentary certificate and known hosts management, so that we can attach host keys or TLS information to repository FQDNs. I'm a little unsure where this would belong, I see two possibilities:

- a dedicated CertificateService with endpoints at e.g. /api/v1/certificate
- an extension of RepositoryService

I think the first would make more sense, since certs/known hosts are valid for multiple repositories, when served from same server.

Third possibilities is one I didn't see yet, but maybe someone else has an idea?

alexec commented on cmd/argocd/commands/repo.go:
```diff
@@ -127,9 +136,9 @@ func NewRepoListCommand(clientOpts *argocdclient.ClientOptions) *cobra.Command {
+			fmt.Fprintf(w, "REPO\tINSECURE\tUSER\tSTATUS\tMESSAGE\n")
 			for _, r := range repos.Items {
-				fmt.Fprintf(w, "%s\t%s\t%s\t%s\n", r.Repo, r.Username, r.ConnectionState.Status, r.ConnectionState.Message)
+				fmt.Fprintf(w, "%s\t%s\t%s\t%s\t%s\n", r.Repo, strconv.FormatBool((r.InsecureIgnoreHostKey || r.Insecure)), r.Username, r.ConnectionState.Status, r.ConnectionState.Message)
```
minor - you can use `%v`

alexec commented on server/repository/repository.proto:
```diff
@@ -44,6 +44,28 @@ message RepoQuery {
+// RepoAccessResponse is the response to RepoAccessQuery
+message RepoAccessResponse {
+	// Whether access is possible
+	bool accessPossible = 1;
```
Are these returned? I can't find mention elsewhere in the PR.

jannfis replied:
Supposed to be used with the next commits - I think it's better to have a dedicated response rather than to display the raw error to the user, even if it'll contain basically the same message later on.

alexec commented on util/git/client.go:
```diff
@@ -57,15 +60,33 @@ func NewFactory() ClientFactory {
+	// explicitly replace it with default client for repositories without the
+	// insecure flag set.
+	if IsHTTPSURL(rawRepoURL) {
+		if insecure {
```
can we always use this library? it's got to work, so lets just always use it

jannfis replied:
Sure, it's basically the counterpart to IsSSHURL() - but as of know, just a wrapper around a regexp's MatchString(), and not thoroughly tested yet.

alexec replied:
I don't think I was clear, I'd like to avoid the if/else here on line 79. I think you can always use the `customHttpClient`?

alexec commented:
Oh - I know I've given a lot of feedback on this, but I think you're doing a splendid job on a tricky problem.

jannfis commented:
> Oh - I know I've given a lot of feedback on this, but I think you're doing a splendid job on a tricky problem.

Your feedback is much appreciated!

jannfis commented:
> I'd like to request changes. It wasn't clear to me when reviewing what `insecure` flags. It appears to combine both flags into one uber flag.
> 
> How we deal with SSH and HTTPS is usually quite independent in the code, and we know even experienced users find it confusing what flags to use.

That's correct, and this is why I chose to simplify both options into one flag where the user just has to state he wants to insecurely connect to the repository, regardless of whether it's SSH or HTTPS connected repository.

> What are the options about having a separate flag named similarly to `kubectl`? E.g. `--insecure-skip-tls-verify`?

It's of course possible, but how things are currently dealt with in the code base regarding repository creation & access, it's just one generic entry point, regardless of whether the repo is SSH or HTTPS. Adding a second flag would create some redundancy without further refactoring the code base, I think.

> 
> I also wonder, how should we document updating the list of known hosts?

Currently, the user would have to edit the ConfigMap `argocd-runtime-cm` and update the `ssh_known_hosts` data, either by using `kubectl edit` or if ArgoCD is self-managed, by editing and pushing the CM's manifest file in the repository. Anyhow, I plan to have a CLI version ready by the time this PR is ready for merging (which obviously will still take some time), maybe even with a fetch cert/host key, verify and store functionality. And of course I will include all accompanying documentation with the PR as well.

alexec commented:
@jessesuen what are your thoughts on have two flags vs having one flag?

jessesuen commented:
> @jessesuen what are your thoughts on having two flags vs having one flag? 

My first reaction was that it was overloading the meaning of field: `Insecure` and I would have preferred two separate fields. I was concerned that there may be other "insecure" flags that users may want to enable/disable separately from host verification. However, after researching it some more, the only insecure options I could find with TLS and SSH were:

* `InsecureSkipVerify`
https://golang.org/pkg/crypto/tls/#Config
* `InsecureIgnoreHostKey`
https://github.com/golang/crypto/blob/master/ssh/client.go#L243

Given that
1. there are no other insecure flags to toggle
2. both of these fields deal with enabling/disabling host verification of the server

I became okay with having a single, context-aware field to control the general concept of insecurely connecting to an unverified host. I don't think a lot of users fully understand the subtleties of when to use `--insecure-skip-tls-verify` vs. `--insecure-ignore-host-key`, so simplifying this into a single option for them may make for a better experience. From a UI perspective, I would similarly present it as a "insecure" checkbox when connecting a repo.

> If we have just one, IMHO we should look to deprecate and remove the redundant flag.

I agree.

alexec commented on cmd/argocd/commands/repo.go:
```diff
@@ -90,7 +98,8 @@ func NewRepoAddCommand(clientOpts *argocdclient.ClientOptions) *cobra.Command {
 	command.Flags().StringVar(&sshPrivateKeyPath, "ssh-private-key-path", "", "path to the private ssh key (e.g. ~/.ssh/id_rsa)")
-	command.Flags().BoolVar(&insecureIgnoreHostKey, "insecure-ignore-host-key", false, "disables SSH strict host key checking")
+	command.Flags().BoolVar(&insecureIgnoreHostKey, "insecure-ignore-host-key", false, "disables SSH strict host key checking (deprecated, use --insecure-skip-server-validation instead)")
+	command.Flags().BoolVar(&insecureSkipServerValidation, "insecure-skip-server-validation", false, "disables server certificate and host key checks")
```
do we want to call this just `--insecure`?

jannfis replied:
The problem is, `--insecure` is already used as a global client flag for connecting insecurely to the ArgoCD API server. How about `--insecure-repository` instead?

alexec commented on manifests/base/config/argocd-runtime-cm.yaml:
```diff
@@ -0,0 +1,16 @@
+  name: argocd-runtime-cm
+data:
+  ssh_known_hosts: |
+    bitbucket.org ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEAubiN81eDcafrgMeLzaFPsw2kNvEcqTKl/VqLat/MaB33pZy0y3rJZtnqwR2qOOvbwKZYKiEO1O6VqNEBxKvJJelCq0dTXWT5pbO2gDXC6h6QDXCaHo6pOHGPUy+YBaGQRGuSusMEASYiWunYN0vCAI8QaXnWMXNMdFP3jHAJH0eDsoiGnLPBlBp4TNm6rYI74nMzgz3B9IikW4WVK+dc8KZJZWYjAuORU3jc1c/NPskD2ASinf8v3xnfXeukU0sJ5N6m5E8VLjObPEO+mN2t/FZTMZLiFqPWc/ALSqnMnnhwrNi2rbfg/rd/IpL8Le3pSBne8+seeFVBoGqzHM9yXw==
```
generated code - no need to check

alexec commented on manifests/namespace-install.yaml:
```diff
@@ -1737,6 +1737,23 @@ metadata:
 apiVersion: v1
+data:
+  ssh_known_hosts: |
+    bitbucket.org ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEAubiN81eDcafrgMeLzaFPsw2kNvEcqTKl/VqLat/MaB33pZy0y3rJZtnqwR2qOOvbwKZYKiEO1O6VqNEBxKvJJelCq0dTXWT5pbO2gDXC6h6QDXCaHo6pOHGPUy+YBaGQRGuSusMEASYiWunYN0vCAI8QaXnWMXNMdFP3jHAJH0eDsoiGnLPBlBp4TNm6rYI74nMzgz3B9IikW4WVK+dc8KZJZWYjAuORU3jc1c/NPskD2ASinf8v3xnfXeukU0sJ5N6m5E8VLjObPEO+mN2t/FZTMZLiFqPWc/ALSqnMnnhwrNi2rbfg/rd/IpL8Le3pSBne8+seeFVBoGqzHM9yXw==
```
generated code

alexec commented on pkg/apis/application/v1alpha1/generated.proto:
```diff
@@ -469,17 +469,25 @@ message ProjectRole {
 
 // Repository is a Git repository holding application configurations
 message Repository {
+  // URL of the repo
```
nice comments

alexec commented on pkg/apis/application/v1alpha1/generated.proto:
```diff
@@ -469,17 +469,25 @@ message ProjectRole {
   optional ConnectionState connectionState = 5;
 
+  // InsecureIgnoreHostKey should not be used anymore, Insecure is favoured
   optional bool insecureIgnoreHostKey = 6;
```
should be annotated with `// DEPRECATED`

alexec commented on pkg/apis/application/v1alpha1/generated.proto:
```diff
@@ -469,17 +469,25 @@ message ProjectRole {
+  // InsecureIgnoreHostKey should not be used anymore, Insecure is favoured
   optional bool insecureIgnoreHostKey = 6;
+
+  // Whether the repo is insecure
```
should state what the implications are

alexec commented on pkg/apis/application/v1alpha1/types.go:
```diff
@@ -844,12 +844,19 @@ type ResourceActionParam struct {
-	InsecureIgnoreHostKey bool            `json:"insecureIgnoreHostKey,omitempty" protobuf:"bytes,6,opt,name=insecureIgnoreHostKey"`
+	// URL of the repo
+	Repo string `json:"repo" protobuf:"bytes,1,opt,name=repo"`
+	// Username for authenticating at the repo server
```
should state this is if HTTPS authentication (not SSH)

alexec commented on pkg/apis/application/v1alpha1/types.go:
```diff
@@ -844,12 +844,19 @@ type ResourceActionParam struct {
+	Repo string `json:"repo" protobuf:"bytes,1,opt,name=repo"`
+	// Username for authenticating at the repo server
+	Username string `json:"username,omitempty" protobuf:"bytes,2,opt,name=username"`
+	// Password for authenticating at the repo server
```
should state this is if HTTPS authentication (not SSH)

alexec commented on pkg/apis/application/v1alpha1/types.go:
```diff
@@ -844,12 +844,19 @@ type ResourceActionParam struct {
+	SSHPrivateKey   string          `json:"sshPrivateKey,omitempty" protobuf:"bytes,4,opt,name=sshPrivateKey"`
+	ConnectionState ConnectionState `json:"connectionState,omitempty" protobuf:"bytes,5,opt,name=connectionState"`
+	// InsecureIgnoreHostKey should not be used anymore, Insecure is favoured
+	InsecureIgnoreHostKey bool `json:"insecureIgnoreHostKey,omitempty" protobuf:"bytes,6,opt,name=insecureIgnoreHostKey"`
```
should be annotated with `// DEPRECATED`

alexec commented on pkg/apis/application/v1alpha1/types.go:
```diff
@@ -862,6 +869,7 @@ func (m *Repository) CopyCredentialsFrom(source *Repository) {
 		m.Password = source.Password
 		m.SSHPrivateKey = source.SSHPrivateKey
 		m.InsecureIgnoreHostKey = source.InsecureIgnoreHostKey
+		m.Insecure = source.Insecure
```
good spot

alexec replied:
is this tested?

alexec commented on server/repository/repository.proto:
```diff
@@ -44,6 +44,28 @@ message RepoQuery {
+	// Password for accessing repo
+	string password = 3;
+	// Private key data for accessing SSH repository
+	string PrivateKey = 4;
```
We call this `SSHPrivateKey` elsewhere. What's with the capitalization of these fields?

alexec commented on util/git/git.go:
```diff
@@ -71,6 +72,11 @@ func IsSSHURL(url string) (bool, string) {
 	return false, ""
 }
 
+// IsSSHURL returns true if supplied URL is HTTPS URL
```
typo in comment

jannfis replied:
Thanks, fixed.

alexec commented on util/settings/settings.go:
```diff
@@ -81,6 +81,7 @@ type RepoCredentials struct {
 	PasswordSecret        *apiv1.SecretKeySelector `json:"passwordSecret,omitempty"`
 	SSHPrivateKeySecret   *apiv1.SecretKeySelector `json:"sshPrivateKeySecret,omitempty"`
 	InsecureIgnoreHostKey bool                     `json:"insecureIgnoreHostKey,omitempty"`
+	Insecure              bool                     `json:"insecure,omitempty"`
```
should this be commented?

jannfis replied:
Yes, you are right. Will fix it with one of the next commits.

alexec commented on util/settings/settings.go:
```diff
@@ -81,6 +81,7 @@ type RepoCredentials struct {
 	PasswordSecret        *apiv1.SecretKeySelector `json:"passwordSecret,omitempty"`
 	SSHPrivateKeySecret   *apiv1.SecretKeySelector `json:"sshPrivateKeySecret,omitempty"`
 	InsecureIgnoreHostKey bool                     `json:"insecureIgnoreHostKey,omitempty"`
```
should this be annotated with `// DEPRECATED`?

jannfis replied:
I tried using it in the comment, but the linter complained heavily about it.

jannfis commented:
Just a little heads up that I've been pretty busy lately but still actively working on this PR. I will come back to your comments soon, @alexec . 

Meanwhile, I have a near feature-complete prototype of an API (and accompanying client functionality) for managing the certificates almost finished.

jannfis commented:
OK, so the latest commits add some large chunk of new stuff. Code still needs tests and polishing as well as user documentation, but these are the main changes and additions:

- Renamed ConfigMap "argocd-runtime-cm" to "argocd-ssh-known-hosts-cm"
- New ConfigMap "argocd-tls-certificates-cm" will hold TLS certificate data for connecting HTTPS repositories
- New API service on the server side: /api/v1/certificates - it supports listing, adding and removing certificates currently. Getting a single certificate for details about it is coming up next.
- New CLI command: cert - it is the client implementation of above mentioned API, and supports all API methods (list, add-ssh, add-tls, rm)
- The Git client in ArgoCD can now use custom CA certificates to verify HTTPS repositories which are either self-signed or signed with a custom CA certificate

jannfis commented:
I have added initial RBAC support to the CertificateService API for Get/Create/Delete operations, however, it is currently all-or-nothing approach - either all certificates, or none. I introduced a new RBAC resource "certificates" for this purpose. 

I am unsure whether it makes sense to extend RBAC functionality to be able to restrict operations on certain certificates, or whether this is an administrative-only operation anyway.

Any thoughts on this?

jannfis commented:
OK, so this PR is almost ready for prime time now. A minor feature is still missing: Get functionality for CertificateService, to retrieve actual certificate data from the server. 

I have written initial user documentation. As I'm not a native English speaker, feel free to correct grammar and spelling. Operator documentation (i.e. handling the new config maps and more internals) is upcoming, I'm unsure where it would fit tho - in "Security", "Declarative Setup" or a dedicated chapter?

Also, there is not a single test yet. Shame on me, gonna have to dig my way in how they work.

jannfis commented:
Build step failing due to lint running out of memory (happened quite often lately - seems CircleCI is a little greedy on RAM)

alexmt commented on util/git/client.go:
```diff
@@ -60,15 +66,24 @@ func NewFactory() ClientFactory {
+	// explicitly replace it with default client for repositories without the
+	// insecure flag set.
+	if IsHTTPSURL(rawRepoURL) {
+		gitclient.InstallProtocol("https", githttp.NewClient(getRepoHTTPClient(rawRepoURL, insecure)))
```
I think  this will override http client for all repos in all git clients. I had to come out with this ugly workaround to solve it: https://github.com/argoproj/argo-cd/pull/1890/files#diff-91bc9daf641d929e791c1b889c32a902

jannfis replied:
Thanks for the hint, I will check for (and explicitly test for) this case. Since I think that every Git repo operation within ArgoCD will use NewClient() with the appropriate parameter set, it should be alright tho. 

getRepoHTTPClient() returns a configured instance of http.Client() according to its repository parametrization, so either with InsecureSkipVerify set to true in the TLS client config, with a CertPool containing CA certs in case the repo has dedicated verification configured or a plain, non-custom one (InsecureSkipVerify is false, system CA cert verification only). So, each Git repository connection should get its own, properly configured http.Client() object.

alexmt replied:
Unfortunately `InstallProtocol` updates global variable which is used by all clients: [plumbing/transport/client/client.go](https://github.com/src-d/go-git/blob/master/plumbing/transport/client/client.go#L16). So it affects all clients, even ones which already exist. Also  `InstallProtocol` updates global map which might cause panic since concurrent map writes/reads are prohibited in golang.

jannfis replied:
I think I understood the issue, and you are right. There is a chance of a race-condition happening, which leads in bad cases to non-accessible repositories (if InstallProtocol() was called a second time with different client settings before the previous client makes the actual connection to Git server), in worst case to panic if InstallProtocol() is called concurrently.

I guess I could integrate your workaround into my PR and replace your `insecureClient` construction with the value returned from the `getRepoHTTPClient()` function in my addition to `util/git/client.go`, because it also supports the addition of custom CA certificates for server verification. Will integrate your additional timeouts along the way.

jannfis replied:
Or maybe the go-git people accept a PR for having a per-client object list of protocols, or at least a single protocol specified.

alexmt replied:
Sure, please feel free to incorporate changes. Most of the code is in https://github.com/argoproj/argo-cd/pull/1890/files#diff-91bc9daf641d929e791c1b889c32a902

I hope so too. Created ticket to add the support: https://github.com/src-d/go-git/issues/1177 .

jannfis replied:
I incorporated & tested your workaround, I just had to change the semantics a little - I don't pass the Insecure option through the Auth objects, because there might be repositories with a NopAuth as well. Instead, I extended the `nativeGitClient` struct to pass that option around (I found that to be much easier).

jannfis commented:
I think this PR is now complete and ready for review. I have added quite a few unit tests, no E2E tests tho. It grew pretty big, so sorry for the massive amount of review work it generates.

There remain a couple of open questions from my side:

- Name of the switch to `repo add` CLI command: `--insecure-skip-server-verification` vs. `--insecure-repository`
 - API endpoint `/api/v1/certificates` ok for everyone?
- Operator documentation, where to put it?

alexec commented on docs/user-guide/private-repositories.md:
````diff
@@ -43,13 +43,90 @@ argocd repo add git@github.com:argoproj/argocd-example-apps.git --ssh-private-ke
+argocd repo add https://git.example.com/test-repo
+```
+
+You can also add more than one PEM for a server by concatenating them into the input stream. This might be useful if the repository server is about to replace the server certificate, possibly with one signed by a different CA. This way, you can have the old (current) as well as the new (future) certificate co-existing. If you already have the old certificate configured, use the `--upsert` flag an ...
````
Neat.

alexec commented on docs/user-guide/private-repositories.md:
```diff
@@ -43,13 +43,90 @@ argocd repo add git@github.com:argoproj/argocd-example-apps.git --ssh-private-ke
 
 ## Self-signed & Untrusted TLS Certificates
 
+> v1.x or higher
```
This is most likely to be v1.2

jannfis replied:
Reflected in last commit

alexec commented on docs/user-guide/private-repositories.md:
```diff
@@ -43,13 +43,90 @@ argocd repo add git@github.com:argoproj/argocd-example-apps.git --ssh-private-ke
+
+Managing TLS certificates via the web UI is currently not possible.
+
+> Before v1.x
```
As above.

jannfis replied:
Reflected in last commit

alexec commented on docs/user-guide/private-repositories.md:
```diff
@@ -43,13 +43,90 @@ argocd repo add git@github.com:argoproj/argocd-example-apps.git --ssh-private-ke
 
 If you are using a privately hosted Git service over SSH, then you have the following  options:
+
+> v1.x or later
```
As above.

jannfis replied:
Reflected in last commit

alexec commented on docs/user-guide/private-repositories.md:
```diff
@@ -43,13 +43,90 @@ argocd repo add git@github.com:argoproj/argocd-example-apps.git --ssh-private-ke
+
+You can also manage SSH known hosts entries in a declarative, self-managed ArgoCD setup. All SSH public host keys are stored in the ConfigMap object `argocd-ssh-known-hosts-cm`.
+
+Managing SSH public host keys via the web UI is currently not possible.
```
Great docs.

alexec commented on docs/user-guide/private-repositories.md:
```diff
@@ -43,13 +43,90 @@ argocd repo add git@github.com:argoproj/argocd-example-apps.git --ssh-private-ke
+
+Managing SSH public host keys via the web UI is currently not possible.
+
+> Before v1.x
```
As above.

jannfis replied:
Reflected in last commit

alexec commented on manifests/base/config/argocd-ssh-known-hosts-cm.yaml:
```diff
@@ -0,0 +1,16 @@
+    app.kubernetes.io/part-of: argocd
+  name: argocd-ssh-known-hosts-cm
+data:
+  ssh_known_hosts: |
```
I checked these once. Should check again.

jannfis replied:
I took the data verbatim from `hack/ssh_known_hosts`, you can probably C&P from the CM data, trim the leading white space, and diff against the file. But that brings me to another question: For declarative setup, it might be benificial to just deliver a CM with blank data, and load the known hosts data initially after first installation, such as
`argocd cert add-ssh --batch --from hack/ssh_known_hosts`
Otherwise I guess, the data in the CM could be overwritten each time a user re-applies the manifest, e.g. at update time. What do you think?

alexec commented on util/db/db.go:
```diff
@@ -37,6 +37,13 @@ type ArgoDB interface {
 	ListHelmRepos(ctx context.Context) ([]*appv1.HelmRepository, error)
+
+	// ListRepoCerticifates lists all configured certificates
+	ListRepoCertificates(ctx context.Context, selector *CertificateListSelector) (*appv1.RepositoryCertificateList, error)
```
minor - but I think we should use `cert` abbreviation throughout the code base

jannfis replied:
You mean like, rename all symbols from `Certificate` to the shorter `Cert`? So in the example of commented section of code, `ListRepoCertificates` would become `ListRepoCerts`, `CertificateListSelector` would become `CertListSelector` and `RepositoryCertificateList` would become `RepositoryCertList` (or even `RepoCertList`) etc etc?

alexec replied:
correct - I think most people would understand what a cert is, and it'll make the code simpler - this is a minor mind you

jannfis replied:
I think this would make the code shorter, but not necessarily simpler :) However, since this would mean quite a lot of refactoring work and I think my tools won't do that automatically, I'd prefer doing this in a separate PR.

alexec replied:
sure, I don't feel strongly about it

alexmt commented on cmd/argocd/commands/repo.go:
```diff
@@ -60,24 +60,32 @@ func NewRepoAddCommand(clientOpts *argocdclient.ClientOptions) *cobra.Command {
 			conn, repoIf := argocdclient.NewClientOrDie(clientOpts).NewRepoClientOrDie()
 			defer util.Close(conn)
+
+			// If the user set a username, but didn't supply password via --password,
```
Looks like we've slightly changed behavior here. The intent was to validate repo and if it is accessible don't even ask username/password. If repo is not accessible and URL looks like https then ask for username/password. 

Can we keep the same behavior but use `repoIf.ValidateAccess` instead of `git.TestRepo`?

jannfis replied:
I think to keep previous behaviour, we would possibly have to send two requests to the server before adding the repository:

First request checks repository access, and on error, we have to evaluate why access was not possible. In case of inability to validate server cert, let the user know and bail out (user has then either to add custom cert, or specify `--insecure-skip-server-verification` for the next run). 

In case of required authorization (then HTTPS connection was successful), prompt the user for credentials and call `repoIf.ValidateAccess` again with the supplied credentials.

However, this would mean to change `repoif.ValidateAccess` to return some meaningful response instead of generic `error`, so that we can differentiate between the reasons for no access. But I think that's a good idea anyway.

jannfis commented on server/repository/repository.proto:
```diff
@@ -94,4 +108,12 @@ service RepositoryService {
 		option (google.api.http).delete = "/api/v1/repositories/{repo}";
 	}
 
+	// ValidateAccessPreCreate validates access to not yet configured repo
```
Oops, you are right. Good spot!

jannfis commented:
Thank you for the reviews and commentary, @alexec, @alexmt  and @jessesuen - it was a pleasure working with you on this PR, and a lot of fun developing it.

From my point of view, you may merge the PR whenever you consider the right time for it. There are a few changes I still want to address (including those mentioned in the above comments), but I think given the already massive scope of this PR, I will address all of those in one or more follow-up PRs.

jannfis commented:
One more bug fix incoming - I wasn't aware that we also use the native Git client on the server side, and unfortunately didn't test it either. Next commit should be the final one, it's running through pre-commit currently.

jannfis commented:
CI failed because of linter runs out of memory (again); passes locally run tests.

jannfis commented:
This requires some more work than expected. So, apologies, I'm working on a good fix for native Git client supporting custom CA bundle.

jannfis commented:
Found a good solution that works perfectly and is not dirty by using `GIT_SSL_CAINFO` environment variable instead of using `-c http.sslCAInfo` command line option. This was well hidden in the documentation of Git. Also, gotta improve my QA processes ;-)

PR now feature complete and hopefully bug-free. Ready for another review or merge, as you see fit.

alexmt commented:
Great, I'm glad you found `GIT_SSL_CAINFO`. Kustomize honors env variable while cloning remote bases, so configured known hosts should work for Kustomize remote bases as well.

Thank you for completing it @jannfis !

jannfis commented:
I think Kustomize remote bases are not covered by the feature-set in this PR, @alexmt - I had a quick look at the code implementing Kustomize in `util/kustomize` and imho, it would need adaption. Current solution is focused on Git repositories connected to ArgoCD. However, I think it'd be not that hard to teach our Kustomize implementation to use the certificates and/or known hosts as used by the Git repository connectors. But I guess that's in scope for a next PR, and I already see a next challenge :)

alexec commented on util/settings/settings.go:
```diff
@@ -80,12 +80,20 @@ type OIDCConfig struct {
-	PasswordSecret        *apiv1.SecretKeySelector `json:"passwordSecret,omitempty"`
-	SSHPrivateKeySecret   *apiv1.SecretKeySelector `json:"sshPrivateKeySecret,omitempty"`
-	InsecureIgnoreHostKey bool                     `json:"insecureIgnoreHostKey,omitempty"`
+	// The URL to the repository
```
good stuff

alexec commented on util/settings/settings.go:
```diff
@@ -196,6 +204,21 @@ func (mgr *SettingsManager) getConfigMap() (*apiv1.ConfigMap, error) {
+// Returns the ConfigMap with the given name from the cluster.
+// The ConfigMap must be labeled with "app.kubernetes.io/part-of: argocd" in
+// order to be retrievable.
+func (mgr *SettingsManager) GetNamedConfigMap(configMapName string) (*apiv1.ConfigMap, error) {
```
minor -  all config map are named - aren't they - how about `GetConfigMapByName`?

jannfis replied:
Much better, clear & concise. Will rename it, thanks!

alexec commented on docs/user-guide/private-repositories.md:
```diff
@@ -43,13 +43,90 @@ argocd repo add git@github.com:argoproj/argocd-example-apps.git --ssh-private-ke
+
+If you are connecting a repository on a HTTPS server using a self-signed certificate, or a certificate signed by a custom Certificate Authority (CA) which are not known to ArgoCD, the repository will not be added due to security reasons. This is indicated by an error message such as `x509: certificate signed by unknown authority`.
+
+1. You can let ArgoCD connect the repository in an insecure way, without verifying the server's certificate at all. This can be accomplished by using the `--insecure-repository` flag when adding the repository with the `argocd` CLI utility. However, this should be done only for non-production setups, as it imposes a serious security issue through possible man-in-the-middle attacks.
```
this is incorrect?

jannfis replied:
I'm unsure about what you mean. What exactly is incorrect?

alexmt replied:
I think @alexec means it supposed to be `insecure-ignore-host-key` instead of `--insecure`

jannfis commented on cmd/argocd/commands/cert.go:
```diff
@@ -0,0 +1,305 @@
+			}
+		},
+	}
+	command.Flags().BoolVar(&removeAllCerts, "remove-all", false, "Remove all configured certificates of all types from server (DANGER: use with care!)")
```
It was convenient during test cycles, but you are right, it's quite a dangerous option (therefore the fat DANGER sign). However, the same could be accomplished by using a `*` wildcard as the repo server name, which is not even really documented. 

I see two options:

1. Ask the user with a prompt if he really wants to perform that dangerous operation when `--remove-all` is specified or `*` was given as the repo-server name without any other selector such as `--cert-type`
2. Discard both possibilities for good, and have the user delete data from ConfigMap objects if they really want to get rid of all certificates.

Thinking of it, sh*t happens too often and I guess variant 2 is the better one. Will remove the code from the CLI. Do you think we should check on server side as well, whether '*' was specified in case someone writes against the API?

jannfis commented on docs/user-guide/private-repositories.md:
````diff
@@ -43,13 +43,90 @@ argocd repo add git@github.com:argoproj/argocd-example-apps.git --ssh-private-ke
+Example for adding a HTTPS repository to ArgoCD without verifying the server's certificate (**Caution:** This is **not** recommended for production use):
+
+```bash
+argocd repo add --insecure-repository https://git.example.com/test-repo
````
In the CLI, it still is `--insecure-skip-server-verification`. But I guess there is still some disagreement on the name of the switch. `--insecure` is already occupied for the connection between CLI and ArgoCD server component. 

But personally I think, when the connection to the repository is insecure, the repository itself should be considered insecure as well - just for the fact that you don't know if ArgoCD is _really_ connecting to the right repo, and not a MITM-installed one. If the flag is `--insecure-repository`, it should be clear to not trust that repository. And to not use it for production purposes.

jannfis commented on server/certificate/certificate.go:
```diff
@@ -0,0 +1,85 @@
+	if err := s.enf.EnforceErr(ctx.Value("claims"), rbacpolicy.ResourceCertificates, rbacpolicy.ActionGet, ""); err != nil {
+		return nil, err
+	}
+	certList, err := s.db.ListRepoCertificates(ctx, &db.CertificateListSelector{
```
Thanks! Will fix.

jannfis commented on server/certificate/certificate.proto:
```diff
@@ -0,0 +1,51 @@
+
+service CertificateService {
+  // List all available certificates
+  rpc List(RepositoryCertificateQuery) returns (github.com.argoproj.argo_cd.pkg.apis.application.v1alpha1.RepositoryCertificateList) {
```
Ah, that is the reason for the swagger going wild. Thanks for the clarification, will rename those methods to the names you proposed, they sound clear & concise to me.

jannfis commented on util/cert/cert.go:
```diff
@@ -0,0 +1,273 @@
+
+	certificateList := make([]string, 0)
+
+	// TODO: Implement maximum amount of data to parse
```
Yes, it's still something that I have to do but deemed not important enough for now. I guess the maximum POST size is already limited by gRPC anyway.

alexmt commented:
Trying to implement e2e tests  for it
